# 01 – Data Overview

## Ziel
Erster Überblick über die TED-Daten: Struktur, Spalten, Datentypen, grundlegende Statistiken.

## Schritte
- Daten laden
- Spalten anzeigen
- Datentypen prüfen
- Fehlende Werte analysieren
- Erste Hypothesen formulieren


Import moduls & style

In [ ]:
# import standard modules
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
from pathlib import Path

# show all columns
pd.set_option('display.max_columns', None)

import sys
from pathlib import Path

### Read data

In [ ]:
%load_ext autoreload
%autoreload 2

import os
import sys

# Main directory
parent_dir = os.path.abspath("..")
if parent_dir not in sys.path:
    sys.path.insert(0, parent_dir)

# Import scripts
from my_scripts.preprocessing import (
    basic_preprocessing,
    filter_germany,
    load_years,
    overview,
)

In [ ]:
# ---------------------------------------------------------
# 1. Settings
# ---------------------------------------------------------

data_path = Path("..\\data")

# select all years
years = list(range(2008, 2017))   # 2008–2016

# select only the necessary columns
usecols = [
    "ISO_COUNTRY_CODE",
    "NUMBER_OFFERS",
    "TYPE_OF_CONTRACT",
    "VALUE_EURO",
    "DT_DISPATCH",
    "CAE_NAME",
    "WIN_COUNTRY_CODE"
]


In [ ]:
# ---------------------------------------------------------
# 2. Downloads for all years
# ---------------------------------------------------------

df = load_years(data_path, years, usecols=usecols)



Total number of rows: 4362869


In [5]:
# ---------------------------------------------------------
# 3. Preprocessing
# ---------------------------------------------------------

df = basic_preprocessing(df)



In [ ]:
overview(df)

In [ ]:
total = len(df)

failed = df[df["NUMBER_OFFERS"].fillna(0) <= 1]
risk = df[df["NUMBER_OFFERS"] == 2]

print("European Union:")
print("Total number of tenders in EU:", len(df))
print("❗ Failed tenders (0–1 participant)):", len(failed))
print("❗ Share (in %) of the total:", round(len(failed) / total * 100, 2), "%")

print("⚠️ High-risk tenders (2 participants):", len(risk))
print("⚠️ Share (in %) of the total:", round(len(risk) / total * 100, 2), "%")

European Union:
❗ Failed tenders (0–1 participant)): 1676668
❗ Share (in %) of the total: 38.43 %
⚠️ High-risk tenders (2 participants): 576874
⚠️ Share (in %) of the total: 13.22 %


### Overall EU Dataset Patterns

The full dataset contains 4.36 million procurement records, providing a broad and diverse basis for statistical analysis and model training.

The share of failed tenders (0–1 participant) is 38.43%, indicating a substantial portion of procedures with insufficient competition.

High‑risk tenders (2 participants) represent 13.22% of all records, suggesting that limited competition is a recurring structural pattern across the EU market.

These proportions highlight consistent cross‑country patterns of low bidder participation, which can be leveraged for robust modeling and comparative analysis.

In [ ]:
# ---------------------------------------------------------
# 4. Filter by Germany only
# ---------------------------------------------------------

df_de = filter_germany(df)

print("Number of tenders in Germany:", len(df_de))



Number of tenders in Germany: 309652


In [ ]:
# ---------------------------------------------------------
# 5. Competition check
# ---------------------------------------------------------

total = len(df_de)

failed = df_de[df_de["NUMBER_OFFERS"].fillna(0) <= 1]
risk = df_de[df_de["NUMBER_OFFERS"] == 2]

print("Germany:")
print("Number of tenders in Germany:", len(df))
print("❗ Failed tenders (0–1 participant)):", len(failed))
print("❗ Share (in %) of the total:", round(len(failed) / total * 100, 2), "%")

print("⚠️ High-risk tenders (2 participants):", len(risk))
print("⚠️ Share (in %) of the total:", round(len(risk) / total * 100, 2), "%")

Germany:
❗ Failed tenders (0–1 participant)): 84747
❗ Share (in %) of the total: 27.37 %
⚠️ High-risk tenders (2 participants): 33235
⚠️ Share (in %) of the total: 10.73 %


### Germany‑Specific Patterns

Germany accounts for 309,652 records, a sufficiently large subset for targeted descriptive analysis but comparatively smaller for standalone predictive modeling.

The share of failed tenders in Germany is 27.37%, notably lower than the EU average, indicating relatively stronger competitive dynamics.

High‑risk tenders represent 10.73%, also below the EU benchmark, suggesting a more stable procurement environment with fewer structurally risky procedures.

The deviation of German indicators from EU‑wide averages implies country‑specific procurement behavior that merits separate examination.